src/SessionSetup/OrbitServiceInstance.cpp

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2022 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "SessionSetup/OrbitServiceInstance.h"
+
+#include <absl/strings/str_format.h>
+
+#include <QCoreApplication>
+#include <QMetaEnum>
+#include <QProcess>
+#include <filesystem>
+#include <memory>
+#include <string>
+
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/Result.h"
+
+namespace orbit_session_setup {
+
+constexpr int kWaitTimeInMilliSeconds = 2000;
+
+class OrbitServiceInstanceImpl : public OrbitServiceInstance {
+ public:
+  explicit OrbitServiceInstanceImpl(const QString& program, const QStringList& arguments);
+  ~OrbitServiceInstanceImpl() override;
+
+  [[nodiscard]] bool IsRunning() const override;
+  [[nodiscard]] ErrorMessageOr<void> Shutdown() override;
+
+  [[nodiscard]] ErrorMessageOr<void> Start();
+
+ private:
+  [[nodiscard]] std::string GetStateErrorAndOutput();
+
+  QProcess process_;
+  QMetaObject::Connection finished_connection_;
+};
+
+ErrorMessageOr<std::unique_ptr<OrbitServiceInstance>> OrbitServiceInstance::Create(
+    const QString& program, const QStringList& arguments) {
+  auto instance = std::make_unique<OrbitServiceInstanceImpl>(program, arguments);
+  OUTCOME_TRY(instance->Start());
+
+  return std::move(instance);
+}
+
+ErrorMessageOr<std::unique_ptr<OrbitServiceInstance>> OrbitServiceInstance::CreatePrivileged() {
+  const QString orbit_service_path = QCoreApplication::applicationDirPath() + "/OrbitService";
+  return OrbitServiceInstance::Create("/usr/bin/pkexec", {orbit_service_path});
```

Comment:
We shouldn't assume pkexec to be in /usr/bin. Does just `pkexec` not work?

Reply:
good idea!